File: mailblastr-python/mailblastr/events.py
"""Events resource — send custom events that trigger automations, plus
custom-event definitions. DOMAIN-FIRST: ``domain`` is REQUIRED on
``Events.send`` (only automations belonging to that domain are triggered)."""

from typing import Any, Dict, TypedDict, Union

from . import http_client
from .http_client import paginate, path_escape as _e


class SendParams(TypedDict, total=False):
    event: str  # the custom event name automations trigger on (`name` = alias)
    name: str  # alias for `event`
    domain: str  # REQUIRED: the sending domain this event belongs to
    contact_id: str  # identify the contact by id (contact_id OR email)
    email: str  # identify the contact by email (contact_id OR email)
    payload: Dict[str, Any]  # arbitrary event payload (`data` = alias)
    data: Dict[str, Any]  # alias for `payload`


class CreateParams(TypedDict, total=False):
    name: str  # required (cannot start with the reserved `mailblastr:` prefix)
    schema: Dict[str, str]  # flat key -> 'string' | 'number' | 'boolean' | 'date'


class UpdateParams(TypedDict, total=False):
    # The name is immutable (automations reference it) — sending `name` is a
    # 422. Only the payload schema can be changed; None clears it.
    schema: Union[Dict[str, str], None]


class Events:
    """``mailblastr.Events`` — the /events endpoints."""

    SendParams = SendParams
    CreateParams = CreateParams
    UpdateParams = UpdateParams

    @classmethod
    def send(cls, params, options=None):
        """Send a custom event that automations can trigger on (``domain`` is
        required). POST /events/send

        ``options={"idempotency_key": ...}`` is still forwarded as
        ``Idempotency-Key``, but only POST /emails and POST /emails/batch
        honour that header — the server ignores it here, so a retry ingests a
        SECOND event and can enroll the contact twice. De-duplicate on your
        side instead."""
        return http_client.request("POST", "/events/send", params, options)

    @classmethod
    def create(cls, params, options=None):
        """Create a custom-event definition (name + optional payload schema).
        POST /events

        A body carrying ``contact_id``/``email`` is treated as an event INGEST
        instead (identical to :meth:`send`). ``options`` carries no idempotency
        guarantee here — see :meth:`send`."""
        return http_client.request("POST", "/events", params, options)

    @classmethod
    def list(cls, params=None):
        """List custom-event definitions. GET /events"""
        return http_client.request("GET", f"/events{paginate(params)}")

    @classmethod
    def update(cls, event_id, params):
        """Update a custom-event definition's payload schema. The event NAME is
        immutable — create a new event to rename. PATCH /events/:id"""
        return http_client.request("PATCH", f"/events/{_e(event_id)}", params)

    @classmethod
    def remove(cls, event_id):
        """Delete a custom-event definition. DELETE /events/:id"""
        return http_client.request("DELETE", f"/events/{_e(event_id)}")

File: mailblastr-ruby/lib/mailblastr/events.rb
# frozen_string_literal: true

module Mailblastr
  # Custom events that automations trigger on. Events are DOMAIN-FIRST:
  # `domain` is required on send, and only that domain's automations fire.
  module Events
    class << self
      # Send a custom event. POST /events/send
      #   Mailblastr::Events.send({ event: "signup.completed", domain: "yourdomain.com",
      #                             email: "user@example.com", payload: { plan: "pro" } })
      # Identify the contact by `contact_id` OR `email`. Event names cannot
      # start with the reserved "mailblastr:" prefix.
      #
      # NOTE: only POST /emails and POST /emails/batch honour `Idempotency-Key`.
      # An `idempotency_key` passed here is still forwarded, but the server
      # ignores it, so a retry ingests a SECOND event and can enroll the contact
      # twice — de-duplicate on your side instead.
      def send(params, options = {})
        Client.require_domain!(params, "Events.send")
        Client.request(:post, "/events/send", body: params, options: options)
      end

      # Create a custom-event definition (name + optional payload schema).
      # Schema values are one of "string", "number", "boolean", "date".
      # POST /events
      #   Mailblastr::Events.create({ name: "signup.completed", schema: { plan: "string" } })
      #
      # NOTE: `options[:idempotency_key]` carries no guarantee here — see
      # `send`. A duplicate event name is already a 422 validation_error.
      def create(params, options = {})
        Client.request(:post, "/events", body: params, options: options)
      end

      # Update a definition's payload schema. PATCH /events/:id
      # The event NAME is immutable (automations reference it) — passing `name`
      # is a 422; pass `schema: nil` to clear the schema.
      #   Mailblastr::Events.update("evt_1", { schema: { plan: "string" } })
      def update(event_id, params)
        Client.request(:patch, "/events/#{Client.path_escape(event_id)}", body: params)
      end

      # List custom-event definitions. GET /events
      def list(params = {})
        Client.request(:get, "/events", query: Client.pagination(params))
      end

      # Delete a custom-event definition. DELETE /events/:id
      def delete(event_id)
        Client.request(:delete, "/events/#{Client.path_escape(event_id)}")
      end
    end
  end
end
